### DISCUSSION TITLE
Subscriptions and persisted queries

### DISCUSSION THREAD

**Discussion by staldera** [enisdenjo/graphql-ws#481] [Q&A] [answered]
Hi,

I currently have a setup where I use Relay + graphql-ws in my frontend and HotChocolate in my backend.
My backend is configured to use persisted queries and block everything else. Thus, it expects `query` to be `null` and the persisted query `id` to be provided.
After setting up Relay to use persisted queries, I compile with the relay compiler, and it also provides an `id`, while setting the query (`text`) field to `null`.

Then graphql-ws comes in, as I'm trying use websockets for the subscriptions. I create a client and try to use it to subscribe:
```
function subscribe(
    operation: RequestParameters,
    variables: Variables,
): Observable<GraphQLResponse> {
    const client = (wsClient ??= createClient({
        url: process.env.REACT_APP_GRAPHQL_WS_ENDPOINT as string,
        retryAttempts: 3
    }));

    return Observable.create((sink) =>
        client.subscribe<GraphQLResponse>(
            {
                operationName: operation.name,
                // query cannot be null for some reason, but accepts an empty string
                query: operation.text ?? "",
                // I'd expect something this to be possible, but id isn't defined in the SubscribePayload type
                // id: operation.id,
                variables,
            },
            {
                next: (response) => sink.next(response as GraphQLResponse),
                complete: sink.complete,
                error: (err) => {
                    if (Array.isArray(err)) {
                        return sink.error(
                            new NetworkError(ErrorMessages.ERROR_FETCH, {
                                cause: err,
                            }),
                        );
                    }

                    if (err instanceof CloseEvent) {
                        return sink.error(
                            new NetworkError(ErrorMessages.SOCKET_CLOSED, {
                                cause: err,
                            }),
                        );
                    }

                    return sink.error(err as any, true);
                },
            },
        ),
    );
}
```

Sadly, `query` can't be `null`, so it behaves poorly on the backend, and `id` can't be provided there, so I have to work arround that too. Is there any reason for this? It seems to me that it's exactly how persisted queries and subscriptions are meant to be used, both from Relay and HotChocolate, but graphql-ws blocks it.
I did some extra tests by changing the payload definition to add the id, and that seemed to work perfectly. However, making the query nullable didn't work.

Edit:
I forget to say, I have also stumbled upon https://the-guild.dev/graphql/ws/recipes#ws-server-and-client-usage-with-persisted-queries
I tried it, and of course I was able to add an extension for the id as well as set the query to an empty string, but as you probably guessed, this doesn't work without rewriting multiple classes in HotChocolate, which I'd tend to avoid.
My biggest problem is the non-nullable `query` field, as fetching the id from extensions and using it to fill the query id in HotChocolate would be doable with an extra middleware (not really the nicest way, but I can live with that).

**Comment by enisdenjo** [comment, marked as answer]
Hey, thanks for writing! The reason `query` is not nullable is because GraphQL over WebSocket spec follows/inherits the [GraphQL over HTTP spec](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#request-parameters) - which also disallows having null for `query` field in request params. Generally wanting to have the `query` field ignored, you'd set an empty string (`{ "query": "" }`).

You're welcome to jump into the [GraphQL over HTTP work group discussion about the optional query field](https://github.com/graphql/graphql-over-http/issues/216).
